### PR TITLE
Add API validation

### PR DESCRIPTION
**Pull request by ketphan02**
- API validation for:
  - `students`
  - `team_generation_options`
  - `algorithm_type`

Since `team_generation_options` is too big, we will have a separate PR for it

### CODE REVIEW

File: api/api/utils/relationship.py [hunk 1]
```diff
@@ -9,3 +9,13 @@ def get_relationship(relationship: str) -> Relationship:
     if relationship == "default":
         return Relationship.DEFAULT
     raise ValueError(f"Relationship {relationship} is not a valid relationship")
+
+
+def get_relationship_str(relationship: Relationship) -> str:
```
Comment (Opeyem1a) [inline api/api/utils/relationship.py:14]: idk if we need this? Why not just `str(relationship)` instead of a util function for it?

File: api/api/utils/generate_teams_data_loader.py [hunk 2]
```diff
@@ -40,10 +40,10 @@ def _get_students(self) -> List[Student]:
         students = self.data.get("students")
         return [
             Student(
```
Comment (Opeyem1a) [inline api/api/utils/generate_teams_data_loader.py:42]: now that I think about it, couldn't we use the serializer here?

File: api/api/validators/generate_teams_validator.py [hunk 3]
```diff
@@ -49,31 +52,148 @@ def validate_algorithm_options(self):
...
-        #
-        #     }
-        # ]).validate(self.data)
+        students: List = self.data.get("students")
+        Schema(
+            [
+                {
+                    "id": int,
+                    Optional("name"): str,
+                    Optional("attributes"): {str: [int]},
```
Comment (Opeyem1a) [inline api/api/validators/generate_teams_validator.py:61]: each of the attribute keys technically need to be able to be parsed into ints, not sure if we can do this with `Schema` but would be cool

File: api/api/validators/generate_teams_validator.py [hunk 4]
```diff
@@ -49,31 +52,148 @@ def validate_algorithm_options(self):
...
+                    Optional("relationships"): {
+                        str: Or(
```
Comment (Opeyem1a) [inline api/api/validators/generate_teams_validator.py:63]: same here with the string technically being an int

File: api/api/utils/relationship.py (hunk 1 above)
Comment (Opeyem1a) [reply api/api/utils/relationship.py:14]: oh because its a float, hmm

File: api/api/validators/generate_teams_validator.py [hunk 5]
```diff
@@ -49,31 +52,148 @@ def validate_algorithm_options(self):
...
+                            ]
+                        )
+                    },
+                    Optional("project_preferences"): [int],
+                }
+            ]
+        ).validate(students)
 
-        pass
+        student_ids = set()
```
Comment (Opeyem1a) [inline api/api/validators/generate_teams_validator.py:75]: for testing purposes as well as readability, I think it might be worth putting each "validation" in a separate function of this class, and just calling them where needed.

i.e. `self.validate_students()` could call `self.validate_student_ids_unique()` and `self.validate_students_project_preferences()`, etc

File: api/api/validators/generate_teams_validator.py [hunk 6]
```diff
@@ -49,31 +52,148 @@ def validate_algorithm_options(self):
...
-        pass
+        team_options = self.data.get("team_generation_options")
+
+        Schema(
+            {
+                "initial_teams": [
+                    {
+                        "id": int,
+                        Optional("name"): str,
+                        "project_id": int,
```
Comment (Opeyem1a) [inline api/api/validators/generate_teams_validator.py:178]: this is also optional

File: api/api/validators/generate_teams_validator.py [hunk 7]
```diff
@@ -49,31 +52,148 @@ def validate_algorithm_options(self):
...
+                        Optional("requirements"): [
+                            {
+                                "attribute": int,
+                                "operator": str,
```
Comment (Opeyem1a) [inline api/api/validators/generate_teams_validator.py:182]: there is a discrete number of possible operators

File: tests/test_api/test_validators/test_generate_teams_validator/test_generate_teams_validator.py [hunk 8]
```diff
@@ -0,0 +1,219 @@
...
+            validator = GenerateTeamsValidator(self.valid_data)
+            validator.validate()
+        except SchemaError as e:
+            self.fail("Valid data raised SchemaError unexpectedly")
+
+    def test_errors__top_levels(self):
+        with self.assertRaises(SchemaError) as e:
+            validator = GenerateTeamsValidator(self.missing_students_data)
+            validator.validate()
+        self.assertEqual(e.exception.code, "Missing key: 'students'")
```
Comment (Opeyem1a) [inline tests/test_api/test_validators/test_generate_teams_validator/test_generate_teams_validator.py:172]: don't remove now, but for the future idk if we should write tests that are this specific, but I know why you did it.

I feel like if you broke the validator class into smaller methods, the tests could also be smaller, and then you wouldnt need to test the exact match of the exception code, you could just test if individual validation methods raise errors

**Review by Opeyem1a — CHANGES_REQUESTED** [review]

File: api/api/utils/generate_teams_data_loader.py (hunk 2 above)
Comment (ketphan02) [reply api/api/utils/generate_teams_data_loader.py:42]: Another PR then

File: api/api/validators/generate_teams_validator.py (hunk 4 above)
Comment (ketphan02) [reply api/api/validators/generate_teams_validator.py:63]: line 136 checked that

File: api/api/validators/generate_teams_validator.py (hunk 3 above)
Comment (ketphan02) [reply api/api/validators/generate_teams_validator.py:61]: line 136 checked that

File: api/api/validators/generate_teams_validator.py [hunk 9]
```diff
@@ -44,32 +50,167 @@ def validate_algorithm_options(self):
...
+                raise SchemaError(f"Student {student.get('id')} has no attributes.")
+
+            # Validate if attribute keys integer string
+            try:
+                attribute_keys = list(map(int, attributes.keys())) if attributes else []
+            except ValueError:
+                raise SchemaError("Attribute keys must be integers.")
+
+            # Validate if attribute keys are unique
+            if len(attribute_keys) != len(set(attribute_keys)):
```
Comment (Opeyem1a) [inline api/api/validators/generate_teams_validator.py:115]: it's impossible to have the same key twice in a dict (hashmap) - my bad if I told you to do this, because on second throught it doesn't need to be checked

File: api/api/validators/generate_teams_validator.py [hunk 10]
```diff
@@ -44,32 +50,167 @@ def validate_algorithm_options(self):
...
+                raise SchemaError("Attribute keys must be unique.")
+
+            # Validate if attribute keys exist
+            if not all_attributes.issuperset(attribute_keys):
```
Comment (Opeyem1a) [inline api/api/validators/generate_teams_validator.py:119]: On second thought here too, it's cool if students have more attribute keys than the project requirements specify, no logical issue there. It's only minorly an issue if project ids have requirements for attributes that no student has, but I don't think we need to check that. It causes no problems for our system, just a wonky team generation for the user (i.e. user error)

tldr; remove

File: api/api/validators/generate_teams_validator.py [hunk 11]
```diff
@@ -44,32 +50,167 @@ def validate_algorithm_options(self):
...
+            relationships = student.get("relationships")
+
+            # Validate if relationship keys integer string
+            try:
+                relationship_keys = list(map(int, relationships.keys()))
+            except ValueError:
+                raise SchemaError("Relationship keys must be integers.")
+
+            # Validate if relationship keys are unique
+            if len(relationship_keys) != len(set(relationship_keys)):
```
Comment (Opeyem1a) [inline api/api/validators/generate_teams_validator.py:137]: unneeded check, keys of dictionaries are unique by definition

**Review by Opeyem1a — CHANGES_REQUESTED** [review]
almost there - just some quick things to remove

**Review by Opeyem1a — APPROVED** [review]
we have awaited this moment 🥺💪
